staging/src/k8s.io/client-go/rest/connection_test.go

```diff
@@ -162,3 +162,72 @@ func TestReconnectBrokenTCP(t *testing.T) {
 		t.Fatalf("expected %d dials, got %d", 2, dials)
 	}
 }
+
+func TestReconnectBrokenTCP_HTTP1(t *testing.T) {
+	ts := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintf(w, "Hello, %s", r.Proto)
+	}))
+	ts.EnableHTTP2 = false
+	ts.StartTLS()
+	defer ts.Close()
+
+	u, err := url.Parse(ts.URL)
+	if err != nil {
+		t.Fatalf("failed to parse URL from %q: %v", ts.URL, err)
+	}
+	lb := newLB(t, u.Host)
+	defer lb.ln.Close()
+	stopCh := make(chan struct{})
+	go lb.serve(stopCh)
+	transport, ok := ts.Client().Transport.(*http.Transport)
+	if !ok {
+		t.Fatalf("failed to assert *http.Transport")
+	}
+	config := &Config{
+		Host:      "https://" + lb.ln.Addr().String(),
+		Transport: utilnet.SetTransportDefaults(transport),
+		Timeout:   1 * time.Second,
+		// These fields are required to create a REST client.
+		ContentConfig: ContentConfig{
+			GroupVersion:         &schema.GroupVersion{},
+			NegotiatedSerializer: &serializer.CodecFactory{},
+		},
+	}
+	config.TLSClientConfig.NextProtos = []string{"http/1.1"}
+	client, err := RESTClientFor(config)
+	if err != nil {
+		t.Fatalf("failed to create REST client: %v", err)
+	}
+	data, err := client.Get().AbsPath("/").DoRaw(context.TODO())
```

Comment:
If the client is doing a WATCH instead of a GET, will the CloseIdleConnections on line 217 close the transport?

Reply:
no, idle in this context doesn't mean "not sending/receiving data", it means "not handling a request and returned to the idle pool"

Reply:
if I understand the scenario correctly this tests does:
1. create an http1.1 connection ,  read, close the body and put it in the pool
2. break the tcp connection
3. try to reuse the connection, that should fail because TCP is broke

we solve it here removing the idle connections from the pool and forcing to create a new connection
(and I understand know Jordan's comment on the next commit :) )

Reply:
CloseIdleConnections() never closes connections that are not in the idle pool.
It turns out that once the request times out it is not returned to the idle pool
https://github.com/kubernetes/kubernetes/pull/104844#discussion_r707220875

Reply:
> once the request times out it is not returned to the idle pool

Hmm... I'm skeptical that is true in all cases, since we saw persistently stuck kubelet -> apiserver connections that just timed out repeatedly ~forever (every few seconds for hours/days) back before http/2 was the default. If a request timing out was sufficient to keep the stuck connection from being returned to the pool, how would the kubelet have remained stuck using broken connections?

Reply:
just trying to find an explanation, can be this before golang 1.7? 1.7 introduced contexts and several important changes on the net/http library, I can see that client-go is using context cancellation for the requests too

Reply:
it was still seen in the k8s 1.8-1.10 timeframe, which was on go 1.8-1.9

https://github.com/kubernetes/kubernetes/blob/release-1.10/hack/lib/golang.sh#L329

Reply:
is it possible that request weren't using timeout until this https://github.com/kubernetes/kubernetes/pull/51042 ?

Reply:
timeout was set client-wide here in the k8s 1.10 timeframe:

https://github.com/kubernetes/kubernetes/blob/f170ef66340f6355d331ed90902574ff0532a20a/staging/src/k8s.io/client-go/rest/config.go#L210-L213

go says that behaves the same as if the context was canceled:

```
	// Timeout specifies a time limit for requests made by this
	// Client. The timeout includes connection time, any
	// redirects, and reading the response body. The timer remains
	// running after Get, Head, Post, or Do return and will
	// interrupt reading of the Response.Body.
	//
	// A Timeout of zero means no timeout.
	//
	// The Client cancels requests to the underlying Transport
	// as if the Request's Context ended.
	//
	// For compatibility, the Client will also use the deprecated
	// CancelRequest method on Transport if found. New
	// RoundTripper implementations should use the Request's Context
	// for cancellation instead of implementing CancelRequest.
	Timeout time.Duration
```

Reply:
yeah, but the client connections weren't timing out until 1.10, no?
let me try to see what happens if I disable the timeout  client side

Reply:
I think that the problem maybe related to pipelining and HOL blocking, I can reproduce it creating more than one connection
```go
func TestReconnectBrokenTCPAfterTimeout_HTTP1(t *testing.T) {
	const parallelRequests = 2
	mux := http.NewServeMux()
	mux.HandleFunc("/slow/", func(w http.ResponseWriter, r *http.Request) {
		// hold the connection
		t.Logf("Request received: %v", r.RemoteAddr)
		time.Sleep(10 * time.Second)
		fmt.Fprintf(w, "Slow Handler, %s", r.Proto)
	})
	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprintf(w, "Hello, %s", r.Proto)
	})
	ts := httptest.NewUnstartedServer(mux)
	ts.EnableHTTP2 = false
	ts.StartTLS()
	defer ts.Close()

	u, err := url.Parse(ts.URL)
	if err != nil {
		t.Fatalf("failed to parse URL from %q: %v", ts.URL, err)
	}
	lb := newLB(t, u.Host)
	defer lb.ln.Close()
	stopCh := make(chan struct{})
	go lb.serve(stopCh)
	transport, ok := ts.Client().Transport.(*http.Transport)
	if !ok {
		t.Fatalf("failed to assert *http.Transport")
	}
	config := &Config{
		Host:      "https://" + lb.ln.Addr().String(),
		Transport: utilnet.SetTransportDefaults(transport),
		Timeout:   3 * time.Second,
		// These fields are required to create a REST client.
		ContentConfig: ContentConfig{
			GroupVersion:         &schema.GroupVersion{},
			NegotiatedSerializer: &serializer.CodecFactory{},
		},
	}
	config.TLSClientConfig.NextProtos = []string{"http/1.1"}
	client, err := RESTClientFor(config)
	if err != nil {
		t.Fatalf("failed to create REST client: %v", err)
	}

	// 1. Create 10 parallel connections
	// 2. Once all of them connect close the TCP connections
	// 3. Wait until the connections timeout
	// 4. Try to connect again, if the connection is not reused it will succeed

	reqConnectCh := make(chan error)
	var once sync.Once
	trace := &httptrace.ClientTrace{
		GotConn: func(connInfo httptrace.GotConnInfo) {
			// t.Logf("Got Conn Reused %v WasIdle %v IdleTime %v", connInfo.Reused, connInfo.WasIdle, connInfo.IdleTime)
		},
		ConnectDone: func(network, addr string, err error) {
			//
		},
		WroteRequest: func(reqInfo httptrace.WroteRequestInfo) {
			// HTTP1.1 pipelining blocks until the first request is received
			once.Do(func() {
				t.Logf("First connection wrote request")
				reqConnectCh <- reqInfo.Err
				close(reqConnectCh)
			})
		},
	}
	ctx := httptrace.WithClientTrace(context.Background(), trace)
	// wait until all the requests times out
	wg := sync.WaitGroup{}
	reqTimeoutCh := make(chan struct{})
	go func() {
		for i := 0; i < parallelRequests; i++ {
			wg.Add(1)
			go func() {
				defer wg.Done()
				_, err := client.Get().AbsPath("/slow").DoRaw(ctx)
				t.Logf("Connection %d finished with err: %v", i, err)
			}()
		}
		wg.Wait()
		fmt.Println("Closing")
		close(reqTimeoutCh)
	}()

	select {
	case <-time.After(10 * time.Second):
		t.Fatalf("Test timed out waiting for first request to fail")
	case err := <-reqConnectCh:
		if err != nil {
			t.Fatalf("Not expected request to fail: %v", err)
		}
	}
	// Deliberately let the LB stop proxying traffic for the current
	// connection. This mimics a broken TCP connection that's not properly
	// closed.
	close(stopCh)

	stopCh = make(chan struct{})
	go lb.serve(stopCh)

	// Wait for the connection to time out
	select {
	case <-reqTimeoutCh:
		//for err := range errCh {
		//	if err == nil {
		//		t.Fatalf("Expected request to timeout")
		//	}
		//	t.Logf("Connection failed with error: %v", err)
		//}
	case <-time.After(10 * time.Second):
		t.Fatalf("Test timed out waiting for first request to fail")
	}

	// Close the idle connections
	utilnet.CloseIdleConnectionsFor(client.Client.Transport)

	// If the client didn't close the idle connections, the broken connection
	// would still be in the connection pool, the following request would
	// then reuse the broken connection instead of creating a new one, and
	// thus would fail.
	t.Logf("Creating new connection after network reconnected")
	data, err := client.Get().AbsPath("/").DoRaw(context.Background())
	if err != nil {
		t.Fatalf("unexpected err: %v", err)
	}
	if string(data) != "Hello, HTTP/1.1" {
		t.Fatalf("unexpected response: %s", data)
	}
	dials := atomic.LoadInt32(&lb.dials)
	if dials != 2 {
		t.Fatalf("expected %d dials, got %d", 2, dials)
	}
}
```

Reply:
I can't reproduce it, the problem of the test was in the custom LB implementation that doesn't accept new connections